src/librustc_typeck/collect/type_of.rs

```diff
@@ -276,14 +287,11 @@ pub(super) fn type_of(tcx: TyCtxt<'_>, def_id: DefId) -> Ty<'_> {
                     let ty = generics
                         .params
                         .iter()
-                        .filter(|param| {
-                            if let ty::GenericParamDefKind::Const = param.kind {
-                                true
-                            } else {
-                                false
-                            }
-                        })
+                        .rev()
                         .nth(arg_index)
+                        // Prevent ourselves from trying to get the type of a lifetime in
+                        // case there are too many generic args.
+                        .filter(|param| matches!(param.kind, ty::GenericParamDefKind::Const))
```

Comment:
When does this occur, considering we're already matching on `arg.id() == hir_id`?

Reply:
Something like
```rust
struct A<'a, const N: usize>(&'a ());

let _: B<4, 7>;
```
afaict we call type_of for `4`, which returns the arg_index `1`. This corresponds to `'a` in the actual struct which causes an ICE later on.

So this should only happen in case there is an invalid number of arguments.

Reply:
Isn't `arg_index` `0` for the lifetime? Sorry, this is all a bit out of cache for me.

Reply:
It normally is, but we index backwards as we don't have to care about `Self` and default params this way.

```
[lifetime, const_arg]
 ^1        , ^0
```

Reply:
I think we want to avoid having logic that depends on the relative order of lifetime, types, consts. For example, if we add default const parameters, or add a new kind of generic parameter, this logic is going to be wrong. Now that I look at it again, even the original code is a bit suspect: it would probably be better not to ignore any arguments, and take the self-type into account.

Reply:
hmm
Not quite sure what to best handle this without depending on argument order.
Have a few ideas that might work though, even if they are more complex

Afaict the current approach is as good as can be, as it only depends on  implicit const params being after the required ones (which is probably always needed). It doesn't work once we add ambiguous generic params though, which is why I ended up changing it.

The longer I think about this the more undesirable ambiguous args become :laughing: 
~~Can't we just require all const args to be prefaced with `const` and be done with it?~~

```rust
struct A<T, const N: usize>;

let _: A<u32, const 7>;
```

Reply:
Well, we can require them to be surrounded by `{}`, which is almost what happens now.

> It doesn't work once we add ambiguous generic params though, which is why I ended up changing it.

We should never have ambiguous parameters, though, only ambiguous arguments. And as soon as we can check the types of the parameters, we can disambiguate. So I wouldn't have thought that would create a problem here?

Reply:
I mean ambiguous args :laughing: 

You probably right though. Afaik Hir is immutable, so I just kept ambiguous args until the end
https://github.com/lcnr/rust/tree/const-param-path
While this could probably be done differently (and better), I am not yet knowledgeable enough to do so :neutral_face:

Reply:
I would need to think about this. Maybe we could use one of the mutable data structures, e.g. like unification does, to record whether arguments are types or consts.

Reply:
I now believe that there are better ways to achieve this.

> Maybe we could use one of the mutable data structures, e.g. like unification does, to record whether arguments are types or consts.

Will get back to you on this once I am more comfortable with the internals and able to have an actually informed opinion.